support loading pointer, structure, array type

ducorduck opened the pull request:
this partially solve [#10](https://github.com/lks9/angr/issues/10#issue-1418408939) to support pointer, structure, array type. I also added an attribute `Compilation_Unit.type_dict` which is a dictionary that map the offset in DWARF to corresponding type object.

ducorduck committed:
support pointer-, array-, structure-type

lks9 committed:
Merge branch 'master' into variableType_update

lks9 commented:
Ok, a few changes I would suggest:

* You have a `CompilationUnit.type_dict`. This means that to access the types one has to navigate to a `cu` object first:
  ```
  type_list = p.loader.main_object.compilation_units[0].type_dict
  ```
  I think it would be better to make `type_list` a global variable of the elf object instead. Since the variable types are defined globally and a variable might reference a type defined in a different cu. And this way, we also save memory because there are more compilation units than elf objects. I also think that `type_list` is a better name than `type_dict`.
* A field in `VariableType` to get the elf object. Then we can also get the `type_list` inside `VariableType` like this:
  ```python
  type_list = self._elf_object.type_list
  ```
  Just as I did with `Variable` which now also has a field `_elf_object`, see 808bf38c827c44abf291f04e40280ab5c1ea6d5a.
* Once we have the elf object in a `VariableType`, we can create nice getter methods:
  ```python
  @attribute
  def pointee(self):
      type_list = self._elf_object.type_list
      return type_list[self.pointee_offset]
  ```
  Likewise for members of a structure:
  ```python
  @attribute
  def members(self):
      type_list = self._elf_object.type_list
      members = []
      for m_offset in self.member_offsets:
          m_type_offset = type_list[m_offset].reference_offset
          members.append( type_list[m_type_offset] )
      return members
  ```
  Or similar, since this does not look good performance-wise, but I think this should work well.
* An array in `C` behaves like a pointer, so it would be nice to have the `VariableArrayType` as subclass of `VariablePointerType`. I also think that `pointee_offset` and `reference_offset` should have the same name (in `VariablePointerType`).

ducorduck committed:
change type_list field, update variable_type.py to dereference offset to type

ducorduck commented:
I've moved `CompilationUnit.type_dict` to 'ELF.type_list', added `variableType._elf_object` and the getter method like you suggested. One small difference is that the getter method would return None if the pointee type is not loaded (because its type is not supported to load yet. ex: typedef, union,..., or some field of the type is missing. Ex: DW_AT_byte_size).

lks9 committed:
Merge branch 'variableType_update' of github.com:lks9/cle into variableType_update

lks9 committed:
Merge branch 'master' into variableType_update

lks9 committed:
Renamed variable types, based on the C standard

lks9 commented:
Ok, I did lots of renaming.

I also had to undo my suggestion of making `ArrayType` a subclass of `PointerType`. I tested it and found out that the `.addr` of an array does not address a pointer (like pointer type) but directly the first element. So they are different. Moreover, the C standard clarly destincts between the *array type* with an *element type* and the *pointer type* with a *referenced type*. I still like `pointee` but I think it is better to go with the C standard and call it `referenced_type`.

lks9 committed:
small typo

lks9 committed:
better structure type presentation
